piet-cairo/src/text.rs

```diff
@@ -210,233 +396,198 @@ impl TextLayout for CairoTextLayout {
     }
 
     fn hit_test_point(&self, point: Point) -> HitTestPoint {
-        // internal logic is using grapheme clusters, but return the text position associated
-        // with the border of the grapheme cluster.
+        let x = (point.x + self.pango_offset.x) * PANGO_SCALE;
+        let y = (point.y + self.pango_offset.y) * PANGO_SCALE;
 
-        // null case
-        if self.text.is_empty() {
-            return HitTestPoint::default();
-        }
-
-        let height = self
-            .line_metrics
-            .last()
-            .map(|lm| lm.y_offset + lm.height)
-            .unwrap_or(0.0);
-
-        // determine whether this click is within the y bounds of the layout,
-        // and what line it coorresponds to. (For points above and below the layout,
-        // we hittest the first and last lines respectively.)
-        let (y_inside, lm) = if point.y < 0. {
-            (false, self.line_metrics.first().unwrap())
-        } else if point.y >= height {
-            (false, self.line_metrics.last().unwrap())
+        let (is_inside, index, trailing) = self.pango_layout.xy_to_index(x as i32, y as i32);
+        let index = if trailing == 0 {
+            index.try_into().unwrap()
         } else {
-            let line = self
-                .line_metrics
-                .iter()
-                .find(|l| point.y >= l.y_offset && point.y < l.y_offset + l.height)
-                .unwrap();
-            (true, line)
+            /*
+             * NOTE(ForLoveOfCats): The docs specify that a non-zero value for trailing
+             * indicates that the point aligns to the trailing edge of the grapheme. In
+             * that case the value tells us the number of "characters" in the grapheme.
+             */
+
+            let text = self.pango_layout.get_text().unwrap();
+            let index = index.try_into().unwrap();
+            let trailing = trailing.try_into().unwrap();
+
+            text[index..]
+                .char_indices()
+                .nth(trailing)
+                .map(|(offset, _)| index + offset)
+                .unwrap_or_else(|| text.len())
```

Comment:
The pango docs aren't super clear on what they mean by a 'character', but I assume it's a codepoint, in which case this looks right. 👍